### PR TITLE
[2.0] Merge ClassMetadataInfo into ClassMetadata

### LINKED ISSUE #1727: [2.0] Merge ClassMetadata and ClassMetadataInfo
Includes deprecation warning for `ClassMetadataInfo` in 1.3.

### PR DESCRIPTION
**Pull request by carusogabriel**
This closes #1727 by merge `ClassMetadataInfo` into `ClassMetadata` as [suggested](https://github.com/doctrine/mongodb-odm/pull/1714#issuecomment-356524003) by @alcaeus in #1714

### CODE REVIEW

File: tests/Doctrine/ODM/MongoDB/Tests/Mapping/ClassMetadataInfoTest.php [hunk 1]
```diff
@@ -5,7 +5,7 @@
 use Doctrine\ODM\MongoDB\DocumentRepository;
 use Doctrine\ODM\MongoDB\Events;
 use Doctrine\ODM\MongoDB\Mapping\Annotations as ODM;
-use Doctrine\ODM\MongoDB\Mapping\ClassMetadataInfo;
+use Doctrine\ODM\MongoDB\Mapping\ClassMetadata;
```
Comment (Majkl578) [inline tests/Doctrine/ODM/MongoDB/Tests/Mapping/ClassMetadataInfoTest.php]: This file and test class should be renamed (or merged with ClassMetadataTest).

Comment (carusogabriel) [reply tests/Doctrine/ODM/MongoDB/Tests/Mapping/ClassMetadataInfoTest.php]: Thanks, this is the conflict, I was tring to find. I’ll merge them as we only have `ClassMetadata` now ☺️

Comment (carusogabriel) [reply tests/Doctrine/ODM/MongoDB/Tests/Mapping/ClassMetadataInfoTest.php]: There's still something going wrong, but is with tests, not the merge :cry:

**Comment by alcaeus** [comment]
I only checked one test (`ClassMetadataTest::testOwningSideAndInverseSide`), but the issue is that the original test called `ClassMetadataInfo::mapManyReference` which doesn't work with the reflection fields and thus didn't fail when a property didn't exist. With the logic combined in one class, the code now errors because we're trying to map a non-existing property. I assume this is the reason for all the failures.

Solution: either create the properties that are causing failing tests or change the failing tests to map existing properties.

**Comment by alcaeus** [comment]
Sure, I'll take a look at them tonight. I'll push changes to your branch 👍

**Comment by alcaeus** [comment]
@carusogabriel looks like I can't push changes for some reason, so I created https://github.com/carusogabriel/mongodb-odm/pull/1.

**Comment by carusogabriel** [comment]
Strange, but okay, merged :smile:

File: lib/Doctrine/ODM/MongoDB/Mapping/Driver/AnnotationDriver.php [hunk 2]
```diff
@@ -32,7 +31,7 @@ class AnnotationDriver extends AbstractAnnotationDriver
      */
     public function loadMetadataForClass($className, ClassMetadata $class)
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Mapping/Driver/AnnotationDriver.php]: Can we use FQCN for Common's `ClassMetadata` and get rid of aliasing our own one?

File: lib/Doctrine/ODM/MongoDB/Mapping/Driver/YamlDriver.php [hunk 3]
```diff
@@ -32,7 +31,7 @@ public function __construct($locator, $fileExtension = self::DEFAULT_FILE_EXTENS
      */
     public function loadMetadataForClass($className, ClassMetadata $class)
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Mapping/Driver/YamlDriver.php]: No need to rename here as this driver is removed in #1734 anyway

File: lib/Doctrine/ODM/MongoDB/Tools/Console/Command/GeneratePersistentCollectionsCommand.php [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 namespace Doctrine\ODM\MongoDB\Tools\Console\Command;
 
-use Doctrine\ODM\MongoDB\Mapping\ClassMetadataInfo;
+use Doctrine\ODM\MongoDB\Mapping\ClassMetadata;
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Tools/Console/Command/GeneratePersistentCollectionsCommand.php]: If I'm seeing correctly this is unused, should be removed

File: lib/Doctrine/ODM/MongoDB/Mapping/Driver/XmlDriver.php [hunk 5]
```diff
@@ -30,7 +29,7 @@ public function __construct($locator, $fileExtension = self::DEFAULT_FILE_EXTENS
      */
     public function loadMetadataForClass($className, ClassMetadata $class)
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Mapping/Driver/XmlDriver.php]: I'd appreciate getting rid of alias here as well

File: lib/Doctrine/ODM/MongoDB/Mapping/Driver/AnnotationDriver.php (hunk 2 above)
Comment (Majkl578) [reply lib/Doctrine/ODM/MongoDB/Mapping/Driver/AnnotationDriver.php]: I'd alias the common one. 😇

File: lib/Doctrine/ODM/MongoDB/Mapping/Driver/YamlDriver.php (hunk 3 above)
Comment (carusogabriel) [reply lib/Doctrine/ODM/MongoDB/Mapping/Driver/YamlDriver.php]: So I'll wait and rebase!

File: lib/Doctrine/ODM/MongoDB/Tools/Console/Command/GeneratePersistentCollectionsCommand.php (hunk 4 above)
Comment (carusogabriel) [reply lib/Doctrine/ODM/MongoDB/Tools/Console/Command/GeneratePersistentCollectionsCommand.php]: I'll see that, thanks!

File: lib/Doctrine/ODM/MongoDB/Mapping/Driver/AnnotationDriver.php (hunk 2 above)
Comment (carusogabriel) [reply lib/Doctrine/ODM/MongoDB/Mapping/Driver/AnnotationDriver.php]: Sorry for being a layman, but what means FQCN? The full namespaces class?

Comment (jmikola) [reply lib/Doctrine/ODM/MongoDB/Mapping/Driver/AnnotationDriver.php]: Fully-Qualified Class Name.

File: lib/Doctrine/ODM/MongoDB/Mapping/Driver/YamlDriver.php (hunk 3 above)
Comment (malarzm) [reply lib/Doctrine/ODM/MongoDB/Mapping/Driver/YamlDriver.php]: No need, I first need to tackle documentation. Please change the other places and let's get this merged, I'll rebase my PR later :)

Comment (malarzm) [reply lib/Doctrine/ODM/MongoDB/Mapping/Driver/YamlDriver.php]: Sorry, you'll need to rebase after all as I've just finished with docs and merged YAML removal ;)

**Comment by carusogabriel** [comment]
The only thing that should be done is the rebase?

**Comment by malarzm** [comment]
Reasbe to solve conflicts and then if you could apply my remarks it'd be great :+1:

**Comment by carusogabriel** [comment]
@malarzm 96f2d7b: you were correct :+1: 

6ff884b: is that you were talking about? I didn't fully understand that :sweat_smile:

**Comment by malarzm** [comment]
Thanks @carusogabriel, sorry for long merging I must have overlooked notification about last change you did :)

**Comment by carusogabriel** [comment]
@malarzm NP, now, let's finish #1714 :construction:

**Pull request merged**